### ISSUE TITLE
cpu BatchNorm 单测有小概率挂掉问题定位与解决方案

### ISSUE THREAD

**Issue by wyg1997** [Oneflow-Inc/oneflow#6526]
### 复现

CI 有非常小的概率在 BatchNorm cpu 后向计算中挂掉：https://github.com/Oneflow-Inc/oneflow/pull/6521/checks?check_run_id=3879653206

可以用以下代码复现：

```python
#!/usr/bin/env python
# coding=utf-8

import numpy as np
import torch

#  import oneflow as torch

bn = torch.nn.BatchNorm1d(3, eps=1e-05, momentum=0.1, affine=False, track_running_stats=False)
bn.train(False)

t = torch.tensor([[[0.3313, 0.3286, 0.3299],
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]]
                )
t.requires_grad_()
res = bn(t)
res.sum().backward()
print(f"origin: {t}")  # 打印出来的结果是一致的
print(f"grad: {t.grad}")  # 梯度不一致
```

梯度分别为：

```
torch: tensor([[[-3.6146e-04,  3.5264e-04,  8.8123e-06],
         [-3.0851e-07,  6.3261e-07, -3.2411e-07],
         [ 6.8978e-07, -1.1698e-07, -5.7280e-07]]])

oneflow: tensor([[[ 7.2292e-04, -7.0529e-04, -1.7630e-05],
         [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
         [ 2.1294e-07, -5.9382e-07, -1.0496e-06]]], dtype=oneflow.float32)
```

可以发现 input.grad 之间有一定的误差，误差最大的有 1e-3 这个数量级。而单测允许的浮点误差为 1e-3，所以单测挂掉了。

### 原因

BatchNorm 的 cpu 前向并没有相应的 functional，其计算是在 python 层用 op 拼出来的，上述例子可以拆成以下的形式：

```python
eps = 1e-05
t = torch.tensor([[[0.3313, 0.3286, 0.3299],
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]]
                )
t.requires_grad_()
mean = t.mean(dim=2, keepdim=True)
var = t.var(dim=2, unbiased=False, keepdim=True)
res = (t - mean) * (var + eps).rsqrt()
res.sum().backward()
print(t.grad)
```

通过对比 numpy 和 oneflow 的结果，可以发现其实 res 的值就有了非常小的误差，大约在 1e-8 这个数量级上。在后向时，这个误差通过一些 sum、乘方差的倒数等计算，被逐渐放大至 1e-4 甚至 1e-3。继续抽取计算过程，可以发现其实在计算均值时（ `mean = t.mean(dim=2, keepdim=True)`），就已经出现了误差。

```python
import numpy as np
import oneflow as flow
import torch

t = torch.tensor([[[0.3313, 0.3286, 0.3299],  # torch
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]]
                )
mean = t.sum(dim=2, keepdim=True) / 3.0

t2 = flow.tensor([[[0.3313, 0.3286, 0.3299],  # oneflow
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]]
                )

mean2 = t2.sum(dim=2, keepdim=True) / 3.0  # mean2 = t2.mean(dim=2, keepdim=True) 在 C++ 层继续拆成先加再除 scalar 的形式

print(f"{mean.numpy() - mean2.numpy()=}")  # oneflow 和 torch 不相等，误差为 2e-8

a = np.array([[[0.3313, 0.3286, 0.3299],
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]]
                ).astype(np.float32)
print(f"{mean.numpy() - a.mean(axis=2, keepdims=True)=}")  # torch 和 numpy 相等
print(f"{mean2.numpy() - a.mean(axis=2, keepdims=True)=}")  # oneflow 和 numpy 不相等
```

其中 `mean2 = t2.sum(dim=2, keepdim=True) / 3.0` 中由于没有 ScalarDiv kernel，在 C++ functor 中的实现为：`ScalarMul(x, Scalar(1.0) / scalar);`，这里 `Scalar(1.0)/scalar` 的结果是 double，但是在 ScalarMul kernel 中变成了 float，这里造成了非常小的误差，最后在 BatchNorm 后向中一点点积累，变成了较大的误差。另外验证：`mean2 = t2.sum(dim=2, keepdim=True) / flow.tensor(3.0)` 来计算均值时，结果是和 torch 一模一样的。

### 解决

这个问题最好的解决方案是直接实现一个 Batchnorm 前后向的 cpu_kernel，避免使用 ScalarDiv functional。

**Comment by hjchen2** [comment]
你上面div_double_factor为什么要先强转成float，直接用double参与计算就行了

**Comment by hjchen2** [comment]
> 1. 误差问题与scalar是Double还是Float无关

所以就是因为scalar被强制转换成了float导致了精度误差，所以和scalar是有关的。

**Comment by hjchen2** [comment]
> houjiang你看下我底下的那个代码实验，用的都是float类型，一个是通过乘法来做除法，一个是原生除法来做，结果就不一样了

嗯，我看到了，但这和我说的并不冲突。浮点值计算本身就是会有精度误差，多个浮点计算会累积误差，这里实际上是把原本一个浮点除法用了一个sclar的浮点除法和一个浮点乘法等价替换，自然比直接做一次浮点除法误差大。但有时候也可能会起到消除误差的作用，比如前一个计算导致了正向的误差，后一个计算导致了负向的误差，累积之后反而误差变小了。

我觉得这个问题并不需要特殊对待，如果非要解决掉这个精度误差，也可以尝试用更高的计算精度，比如double。

**Comment by wyg1997** [comment]
> > > 1. 误差问题与scalar是Double还是Float无关
> > 
> > 
> > 就是因为scalar被强制转换成了float导致了精度误差，所以和scalar是有关的。
> 
> houjiang你看下我底下的那个代码实验，用的都是float类型，一个是通过乘法来做除法，一个是原生除法来做，结果就不一样了

上面的例子中如果不是 a + b + c，换成 a + b 或 a + c 或者其它确定的数，乘法和除倒数结果就会是一致的了。但实际中的例子中，tensor.sum 的结果是一致的，这里可能是 C++ 表示 0.9898 这个浮点数有些特殊的问题。

**Comment by daquexian** [comment]
实现一个正常的 scalar_div（而不是用 scalar_mul 模拟）还是需要的吧，少一个误差来源总比多一个误差来源好

**Comment by hjchen2** [comment]
> 实现一个正常的 scalar_div（而不是用 scalar_mul 模拟）还是需要的吧，少一个误差来源总比多一个误差来源好

也可以～

**Comment by MARD1NO** [comment]
@simonJJJ shijie最近的重构部分，也对这个scalar div添加一个functor和一些修改吧

**Comment by wyg1997** [comment]
> 实现一个正常的 scalar_div（而不是用 scalar_mul 模拟）还是需要的吧，少一个误差来源总比多一个误差来源好

实现这个的话，在什么时候会用到呢？如果直接把 `tensor * (1.0 / scalar)` 这个模拟替掉的话，速度上会慢不少

**Comment by yuanms2** [comment]
https://github.com/Oneflow-Inc/oneflow/commit/e7af063827f3b0ef95e0ced9899ecb9057e023d7

这个PR 是否解决了这个问题

**Comment by wyg1997** [comment]
> [e7af063](https://github.com/Oneflow-Inc/oneflow/commit/e7af063827f3b0ef95e0ced9899ecb9057e023d7)
> 
> 这个PR 是否解决了这个问题

上午验证了一下，后面两个通道的值一致了，第一个通道还有误差，还需要再 debug 一下

**Comment by Ldpe2G** [comment]
关于第一个通道，存在 oneflow 梯度是 torch -2 倍的关系，做了一个简单的实现，就是 torch cuda 与 cpu 的对比：
```python3
import numpy as np
import torch

device="cuda"
bn = torch.nn.BatchNorm1d(3, eps=1e-05, momentum=0.1, affine=False, track_running_stats=False)
bn.train(False)
bn.to(device)

t = torch.tensor([[[0.3313, 0.3286, 0.3299],
                   [0.6543, 0.1114, 0.6633],
                   [0.3048, 0.5179, 0.6383]]],device=device
                )
t.requires_grad_()
res = bn(t)
res.sum().backward()
print(f"input grad: {t.grad}")
```


```
# torch cpu input grad
input grad: tensor([[[-3.6146e-04,  3.5264e-04,  8.8123e-06],
         [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
         [ 6.8978e-07, -1.1698e-07, -5.7280e-07]]])
# torch cuda input grad
input grad: tensor([[[ 7.1192e-04, -7.1192e-04, -1.7798e-05],
         [ 0.0000e+00,  0.0000e+00,  0.0000e+00],
         [ 8.6428e-07,  0.0000e+00, -4.3214e-07]]], device='cuda:0')
```

torch 也存在 cuda 梯度是 cpu 实现梯度 -2 倍的关系

**Comment by wyg1997** [comment]
已实现 cpu batchnorm kernel，此问题不存在，关闭 issue
